### PR TITLE
[Model][Pytorch] GraphSAGE

### PR DESCRIPTION
**Pull request by hbsun2113**
## Description
simple implemention of GraphSAGE. 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: examples/pytorch/graphsage/graphsage.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+        if aggregator_type == "pooling":
+            self.aggregator = PoolingAggregator(g, in_feats, out_feats, activation, bias)
+        else:
+            self.aggregator = MeanAggregator(g, in_feats, out_feats, activation, bias)
+
+    def forward(self, h):
+        if self.dropout:
+            h = self.dropout(h)
+        self.g.ndata['h'] = h
+        self.g.update_all(graphsage_msg, self.aggregator)
```
Comment (ZiyueHuang) [inline examples/pytorch/graphsage/graphsage.py]: You can use the built-in message function `fn.copy_src(src='h', out='m')`

File: examples/pytorch/graphsage/graphsage.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        h = self.concat(h, nei, node)  # (B, F)
+        h = torch.mm(h, self.weight)  # (B, EF)
+        if self.bias is not None:
+            h = h + self.bias
+        if self.activation:
+            h = self.activation(h)
+        return {'h': h}
+
+    def concat(self, h, nei, nodes):
+        print('no implementation !')
```
Comment (ZiyueHuang) [inline examples/pytorch/graphsage/graphsage.py]: raise error here?

**Comment by ZiyueHuang** [comment]
Do you have any plan to implement other aggregators such as LSTM aggregator?

File: examples/pytorch/graphsage/graphsage.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+class Aggregator(nn.Module):
+    def __init__(self, g, in_feats, out_feats, activation=None, bias=True):
+        super(Aggregator, self).__init__()
+        self.g = g
+        self.weight = nn.Parameter(torch.Tensor(in_feats, out_feats))  # (F,EF)
+        if bias:
+            self.bias = nn.Parameter(torch.Tensor(out_feats))  # (EF,1)
+        else:
+            self.bias = None
```
Comment (BarclayII) [inline examples/pytorch/graphsage/graphsage.py]: L25-L29: You can simply use `nn.Linear` instead, like this:
```python
    self.linear = nn.Linear(in_feats, out_feats, bias=bias)
```

File: examples/pytorch/graphsage/graphsage.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+        self.activation = activation
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        weight_stdv = 1. / math.sqrt(self.weight.size(1))
+        self.weight.data.uniform_(-weight_stdv, weight_stdv)
+        if self.bias is not None:
+            bias_stdv = 1. / math.sqrt(self.bias.size(0))
+            self.bias.data.uniform_(-bias_stdv, bias_stdv)
```
Comment (BarclayII) [inline examples/pytorch/graphsage/graphsage.py]: This procedure can be replaced with `nn.init`.  For example, if you are using ReLU activations, the following runs Xavier uniform initialization with a correction:
```python
        nn.init.xavier_uniform_(self.linear.weight, gain=nn.init.calculate_gain('relu'))
        # if you stick with your code with individual parameters that will be self.weight.
```

Only a hint, since your initialization is different from Xavier uniform.  I saw the original implementation using Xavier uniform though.

File: examples/pytorch/graphsage/graphsage.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+    def forward(self, node):
+        nei = node.mailbox['m']  # (B, N, F)
+        h = node.data['h']  # (B, F)
+        h = self.concat(h, nei, node)  # (B, F)
+        h = torch.mm(h, self.weight)  # (B, EF)
+        if self.bias is not None:
+            h = h + self.bias
```
Comment (BarclayII) [inline examples/pytorch/graphsage/graphsage.py]: L44-46 can then be replaced with
```python
        h = self.linear(h)
```

File: examples/pytorch/graphsage/graphsage.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        if self.activation:
+            h = self.activation(h)
+        return {'h': h}
+
+    def concat(self, h, nei, nodes):
+        print('no implementation !')
+        pass
```
Comment (BarclayII) [inline examples/pytorch/graphsage/graphsage.py]:
```python
        raise NotImplementedError
```
Also remove the `pass` statement.

File: examples/pytorch/graphsage/graphsage.py (hunk 1 above)
Comment (hbsun2113) [reply examples/pytorch/graphsage/graphsage.py]: Thanks for your reivew! I got it!

File: examples/pytorch/graphsage/graphsage.py (hunk 2 above)
Comment (hbsun2113) [reply examples/pytorch/graphsage/graphsage.py]: No, it is just an abstract method.

File: examples/pytorch/graphsage/graphsage.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+    def __init__(self, g, in_feats, out_feats, activation, bias):  # (2F, F)
+        super(PoolingAggregator, self).__init__(g, in_feats*2, out_feats, activation, bias)
+        self.mlp = PoolingAggregator.MLP(in_feats, in_feats, F.relu, False, True)
+
+    def concat(self, h, nei, nodes):
+        nei = self.mlp(nei)
+        concatenate = torch.cat((nei, h), 1)
+        return concatenate
+
+    class MLP(nn.Module):
```
Comment (BarclayII) [inline examples/pytorch/graphsage/graphsage.py]: Move this outside the `PoolingAggregator` class.

File: examples/pytorch/graphsage/graphsage.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+        def __init__(self, in_feats, out_feats, activation, dropout, bias):  # (F, F)
+            super(PoolingAggregator.MLP, self).__init__()
+            self.weight = nn.Parameter(torch.Tensor(in_feats, out_feats))  # (F,EF)
+            if dropout:
+                self.dropout = nn.Dropout(p=dropout)
+            else:
+                self.dropout = 0.
```
Comment (BarclayII) [inline examples/pytorch/graphsage/graphsage.py]: L83-86: just do
```python
self.dropout = nn.Dropout(p=dropout)
```
`dropout = 0` automatically disables it, so you can always pass the input to the dropout layer regardless.

File: examples/pytorch/graphsage/graphsage.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+            if bias:
+                self.bias = nn.Parameter(torch.Tensor(out_feats))  # (EF)
+            else:
+                self.bias = None
```
Comment (BarclayII) [inline examples/pytorch/graphsage/graphsage.py]: L82, L87-L90: similarly you can just do
```python
self.linear = nn.Linear(in_feats, out_feats, bias=bias)
```

**Comment by hbsun2113** [comment]
> Do you have any plan to implement other aggregators such as LSTM aggregator?

There are three aggregators in the original paper, and I have implemented two of them.

File: examples/pytorch/graphsage/graphsage.py [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+            self.activation = activation
+            self.reset_parameters()
+
+        def reset_parameters(self):
+            weight_stdv = 1. / math.sqrt(self.weight.size(1))
+            self.weight.data.uniform_(-weight_stdv, weight_stdv)
+            if self.bias is not None:
+                bias_stdv = 1. / math.sqrt(self.bias.size(0))
+                self.bias.data.uniform_(-bias_stdv, bias_stdv)
```
Comment (BarclayII) [inline examples/pytorch/graphsage/graphsage.py]: Again, consider `torch.nn.init.xavier_uniform_` in this function.

File: examples/pytorch/graphsage/graphsage.py [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+
+        def forward(self, nei):  # (B, N, F)
+            if self.dropout:
+                nei = self.dropout(nei)
```
Comment (BarclayII) [inline examples/pytorch/graphsage/graphsage.py]: Considering the changes in L83-L86, you can simply always do
```python
nei = self.dropout(nei)
```

File: examples/pytorch/graphsage/graphsage.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+            nei = torch.matmul(nei, self.weight)  # (B, N, EF)
+            if self.bias is not None:
+                nei = nei + self.bias
```
Comment (BarclayII) [inline examples/pytorch/graphsage/graphsage.py]: L104-106:
```python
nei = self.linear(nei)
```

File: examples/pytorch/graphsage/graphsage.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+                 dropout,
+                 aggregator_type,
+                 bias=True,
+                 ):
+        super(GraphSAGELayer, self).__init__()
+        self.g = g
+        if dropout:
+            self.dropout = nn.Dropout(p=dropout)
+        else:
+            self.dropout = 0.
```
Comment (BarclayII) [inline examples/pytorch/graphsage/graphsage.py]: Again, just do `self.dropout = nn.Dropout(p=dropout)`

File: examples/pytorch/graphsage/graphsage.py (hunk 1 above)
Comment (BarclayII) [reply examples/pytorch/graphsage/graphsage.py]: Meaning: replace `graphsage_msg` with `fn.copy_src(src='h', out='m')`.

File: examples/pytorch/graphsage/graphsage.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+        if aggregator_type == "pooling":
+            self.aggregator = PoolingAggregator(g, in_feats, out_feats, activation, bias)
+        else:
+            self.aggregator = MeanAggregator(g, in_feats, out_feats, activation, bias)
+
+    def forward(self, h):
+        if self.dropout:
+            h = self.dropout(h)
```
Comment (BarclayII) [inline examples/pytorch/graphsage/graphsage.py]: Again, always do `h = self.dropout(h)`.

File: examples/pytorch/graphsage/graphsage.py (hunk 1 above)
Comment (hbsun2113) [reply examples/pytorch/graphsage/graphsage.py]: What is the difference between them?

File: examples/pytorch/graphsage/graphsage.py (hunk 2 above)
Comment (hbsun2113) [reply examples/pytorch/graphsage/graphsage.py]: I fix it by:
```
import abc
@abc.abstractmethod
    def concat(self, h, nei, nodes):
        raise NotImplementedError
```

File: examples/pytorch/graphsage/graphsage.py (hunk 1 above)
Comment (BarclayII) [reply examples/pytorch/graphsage/graphsage.py]: There is no difference semantically.  However,
* You write one less Python function,
* It allows possibility for DGL to further optimize on speed, since you are using a DGL-provided builtin function.

File: examples/pytorch/graphsage/graphsage.py (hunk 7 above)
Comment (hbsun2113) [reply examples/pytorch/graphsage/graphsage.py]: Thanks for your review!! I learned a lot!
 I wonder why it is better?

**Comment by hbsun2113** [comment]
@BarclayII @ZiyueHuang 
Thanks for your review! I have fixed the code as you suggest. Can you tell me how to update the pull request?

File: examples/pytorch/graphsage/graphsage.py [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+import argparse, time, math
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from dgl import DGLGraph
+from dgl.data import register_data_args, load_data
+
+
+def graphsage_msg(edge):
```
Comment (hbsun2113) [inline examples/pytorch/graphsage/graphsage.py]: test

**Comment by jermainewang** [comment]
> Thanks for your review! I have fixed the code as you suggest. Can you tell me how to update the pull request?

@hbsun2113 Simply push commits normally to your `graphsage` branch and the PR will be updated accordingly.

File: examples/pytorch/graphsage/graphsage.py (hunk 8 above)
Comment (jermainewang) [reply examples/pytorch/graphsage/graphsage.py]: TIL

File: examples/pytorch/graphsage/graphsage.py (hunk 7 above)
Comment (jermainewang) [reply examples/pytorch/graphsage/graphsage.py]: Nested/Inner class should usually be avoided. To me, there is no benefit. In terms of encapsulation, I could still access the inner class as a class attribute like `PoolingAggregator.MLP`. So why not just `MLP`? Usually when I saw a nested class I will try to think "hmm... why is it needed? is there any magic?", but the answer is usually no.

File: examples/pytorch/graphsage/graphsage.py (hunk 8 above)
Comment (hbsun2113) [reply examples/pytorch/graphsage/graphsage.py]: > TIL

What is the meaning?

**Comment by hbsun2113** [comment]
I have updated the code as @BarclayII and @ZiyueHuang suggest. Thx!

**Comment by jermainewang** [comment]
@BarclayII any further comments?

**Comment by jermainewang** [comment]
Have added you to the contributor list.

**Comment by BarclayII** [comment]
The MLP class is still nested.  No big deal though.

Thanks again for your contribution!
